infrastructure/README.md

```diff
@@ -0,0 +1,132 @@
+
+# Goals & Non-Goals
+
+ - Be Language-Agnostic (does not work "better" with any particular platform)
+ - Be Portable (minimal platform dependency)
+ - Be Clear (all server management actions are a directly infered from the source)
+ - Be Declarative (can rollback source to rollback server)
+ - Be Siloed (all dependencies and resources of an app are separated from others)
+ - Be Easy (if we want this to stick, its got to be easy to use)
+
+# Design
+
+## Contain Yourself
+
+`Docker` is an easy solution to the silo and portable bullets (although its not very declarative or clear),
+so we will use it to contain every app that we want to run.
+It doesn't matter that `Docker` isn't declarative since we don't care if the actual apps are.
+
+I know, maybe some of you wanted `lxce` or `nixOS`, I'm sorry.
+
+## Freight
+
+We will be using Kubernetes, since its gives us an easy way to run a an arbitrary amount of containers
+at once.
+
+## The App Repo
+
+Often a design can be gleaned from thinking of "what we would like to happen".
+So if I'm writing an app for a hackathon I would like it to be portable across all events,
+and in doing so I do not want to be responsible for placing this app in any specific event.
+
+So the steps to writing an app would be:
+
+1. Write your damn app
+2. Create a [Dockerfile](https://docs.docker.com/engine/reference/builder/) that runs your app.
+3. Create a `deployment.yaml` file which specifies what resources your app needs and
+   what other containers (if any) it depends on. This will be equivalent to the `spec.containers`
+   section of a [K8s deployment configuration](http://kubernetes.io/docs/user-guide/deployments/).
+
+Now when building or running the docker image the developer will have access to the entire event metadata
+(described later) in the form of JSON in an environment variable:
+
+```bash
+HACKATHON_METADATA='{"name":"HackGT","domain":"hack.gt", ...}'
+```
+
+This is helpful if you want to display the name of the event, etc. in your app.
+
+### Tests
+
+Every Dockerfile that matches `/tests/Dockerfile` will be built and ran during testing.
+For example: if you have a websocket component of your app in `<git root>/websockets`
+you can write tests for it and create a container to run them in `websockets/tests/`.
+By doing this you create a Dockerfile in `websockets/tests/Dockerfile`, which matches.
+
+## The Event Repo
+
+When we know what event we are running these apps on we get to know a lot more,
+this is a simple repo that contains metadata about each event:
+
+```yaml
+name: "HackGT"
+times:
+  - start: 12345
+  - end: 12345
+  - submission: 12345
+homepage: "https://hack.gt/"
+domain: "hack.gt"
+apps:
+  hackgt/helpq:
+    - branch: master # this is default
+    - url: helpmewith.hack.gt
+  hackgt/eventbriter:
+    - url: tickets.hack.gt
+  etc..
+```
+
+This metadata gets its own repository and is given a special name: `gaggle.yaml`.
```

Comment:
> This metadata gets its own repository...

This very briefly lost me. I think slightly more explanation here might be worthwhile (or included in the introduction as a high-level roadmap).

To clarify, applications have their own repositories. Then for each event, we have make a new repository (ex. `hackgteeny17` repo) with this `gaggle.yaml` (please rename) file. It specifies which apps will be run and from where (git repo url).